### ISSUE TITLE
[Feature Request]: string validation with regex

### ISSUE THREAD

**Issue by greyblake** [greyblake/nutype#5]
It was requested a few times on reddit and also stays in my [TODO/WORKLOG](https://github.com/greyblake/nutype/blob/master/WORKLOG.md).

## Syntax

```rs
#[nutype(
    validate(
        regex = "^\w+@\w(\.\w)+$"
    )
)]
pub struct Username(String);
```

Alternative syntax could be with `//` like in scripting languages:

```rust
regex = /^\w+@\w(\.\w)+$/
```

If you're reading this, please share your opinion.

## Crates

* [regex](https://crates.io/crates/regex)
* [lazy-regex](https://crates.io/crates/lazy-regex/) - seems to be based on top of `regex`. Can be very helpful, since the regex is check at compile time

## Concerns

Will it play well with potential support of Arbitrary?
I think [rand_regex](https://docs.rs/rand_regex/latest/rand_regex/index.html) can be helpful here.

**Comment by jssblck** [comment]
> Alternative syntax could be with // like in scripting languages:
> If you're reading this, pleasure share your opinion.

As a user, I would prefer something more like the first option:
```rust
#[nutype(
    validate(
        regex = r#"^\w+@\w(\.\w)+$"#
    )
)]
pub struct Username(String);
```

This is because I'd rather keep magic in macros to a minimum, keeping the code as "rust-like" as possible. I think this is easier for the library as well: just accept a string here and leave it up to the user to create.

**Comment by tyranron** [comment]
> Alternative syntax could be with `//` like in scripting languages:
> 
> ```rust
> regex = /^\w+@\w(\.\w)+$/
> ```
> 
> If you're reading this, pleasure share your opinion.

I'd prefer seeing macro input the same as `regex` crate's input. So, definitely a string literal.

> * [lazy-regex](https://crates.io/crates/lazy-regex/) - seems to be based on top of `regex`. Can be very helpful, since the regex is check at compile time

No need to increase number of dependencies for that. Regex may be trivially validated by proc macro expanding it.

It would also be nice to accept not only string literals, but constants and constant expressions too. Sometimes we do want to keep regex description separately and reuse in multiple places, despite the fact that in such cases we won't be able to validate them in compile-time.

**Comment by greyblake** [comment]
@tyranron @jssblck Thanks to your contribution.

Ok, so we agree we don't want `regex = /abc/`, but rather 

`regex = "abc"` and `regex = r#"abc"#` (both are just normal strings  from the rustacean point of view).

> No need to increase number of dependencies for that. Regex may be trivially validated by proc macro expanding it.

This is a good point. A tricky part is: if the regex is validated at the moment of expanding it means:
* `regex` dependency is needed at compile time
* `regex` versions by at compile time and runtime may diverge, which may result into _oopsy_ behaviour (e.g. when something compiles but then suddenly crashes at runtime).

> It would also be nice to accept not only string literals, but constants and constant expressions too

I agree, though don't have any clear idea yet on that. Shell the constant by of type `&'static str` or `Regex` or something else ?

For example I often use `Regex` in pair with `once_cell` (earlier it was `lazy_static`) to do something like this:

```rs
static SOCKET_REGEX: Lazy<Regex> = Lazy::new(|| {
    Regex::new(
        r#"(?x)
        [^\d](\d{1,3}\.\d{1,3}.\d{1,3}.\d{1,3}:\d{2,5})[^\d]
    "#,
    )
    .unwrap()
});
```

In this case doing something like 

```rs
#[nutype(
    validate(regex = SOCKET_REGEX)
)]
pub struct Socket(String);
```

is tricky, cause `nutype` has no chance to know the type of `SOCKET_REGEX` at the moment of macro expanding.

What are yours development habits?

**Comment by tyranron** [comment]
> A tricky part is: if the regex is validated at the moment of expanding it means:
> 
>     * `regex` dependency is needed at compile time
> 
>     * `regex` versions by at compile time and runtime may diverge, which may result into _oopsy_ behaviour (e.g. when something compiles but then suddenly crashes at runtime).

No, it's quite trivial, imho. `regex` dependency is needed in both at compile time and at runtime, so you use the same version in both `nutype` and `nutype_macros` crates, which is trivially controllable. In `nutype_macros` crate, you use it for validating string literals. In `nutype` crate you keep its re-export, so in proc macro expansion you use something like:
```rust
static __NUTYPE_REGEX: ::nutype::once_cell::sync::Lazy<::nutype::regex::Regex> = 
    ::nutype::once_cell::sync::Lazy::new(|| {
        :nutype::regex::Regex::new(#regex_lit).unwrap()
    })
```

> I agree, though don't have any clear idea yet on that. Shell the constant by of type `&'static str` or `Regex` or something else ?

First, I was thinking just about any expression returning `&'static str`.

> ```rust
> #[nutype(
>     validate(regex = SOCKET_REGEX)
> )]
> pub struct Socket(String);
> ```

But being able to feed there `Regex` or `Lazy<Regex>` is a very nice idea!

> is tricky, cause `nutype` has no chance to know the type of `SOCKET_REGEX` at the moment of macro expanding.

This is easily solvable by some additional type-level glue in `nutype` crate, which will aid the macro expansion with return-type polymorphism:
```rust
// in `nutype` crate

pub trait IntoRegex: Sized {
    fn into_regex(self) -> Lazy<Regex>
}

impl IntoRegex for &str {
    fn into_regex(self) -> Lazy<Regex> {
        Lazy::new(|| Regex::new(self).unwrap())
    }
}

impl IntoRegex for Regex {
    fn into_regex(self) -> Lazy<Self> {
        Lazy::new(|| self)
    }
}

impl IntoRegex for Lazy<Regex> {
    fn into_regex(self) -> Self {
        self
    }
}
```
and expand it like:
```rust
static __NUTYPE_REGEX: ::nutype::once_cell::sync::Lazy<::nutype::regex::Regex> = 
    ::nutype::glue::IntoRegex::into_regex(#whatever)
```

**Comment by greyblake** [comment]
> No, it's quite trivial, imho. regex dependency is needed in both at compile time and at runtime, so you use the same version in both nutype and nutype_macros crates, which is trivially controllable. 

I am not convinced that it will always be true. There can be the following deps tree:

* App
  * regex **version A**
  * nutype
    * nutype_macros
      * regex **version B**

However, maybe it's good enough to go, assume that in most cases regex version will be the same and in case of differences, the differences won't cause any problems (hopefully), cause regex syntax is very stable.

> This is easily solvable by some additional type-level glue
Good point, I forgot that we can play that trick.

Although, some users may not use `once_cell` so it may require one more extra feature flag.


@tyranron @jssblck What do you think would be a proper error variant when a regex is violated? 
My initial idea was to have an error variant per validator. At this point `Invalid` is already used custom validator `with`.
I feel like `Invalid` could be a good name, but then it breaks this 1-1 match between validators and error variants.

**Comment by tyranron** [comment]
@greyblake `RegexMismatch`/`ed`?

**Comment by greyblake** [comment]
InvalidFormat?

**Comment by greyblake** [comment]
@jssblck @tyranron Addressed in https://github.com/greyblake/nutype/pull/24.

Hopefully I will release v0.2.0 next week with this included.

**Comment by greyblake** [comment]
@jssblck @tyranron   FYI the feature is available in 0.2.0.
